✨ feat: 강의 북마크 추가,삭제 API 작성

Kochiya35 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #40
- 작업 요약: 강의 북마크 추가or삭제 기능

## 📄 상세 내용
- [ 강의 북마크 추가or삭제 기능 ] 
- [ managers.py 생성 ]

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Kochiya35 committed:
✨ feat: 강의 북마크 토글 API 작성
REQ-LECT-005 강의 북마크 추가 or 삭제 및 테스트

# Conflicts:
#	apps/lectures/migrations/0004_alter_category_created_at_alter_category_updated_at_and_more.py

SimJunSik commented on apps/lectures/views/bookmark_views.py:
```diff
@@ -0,0 +1,55 @@
+        404: {"type": "object", "properties": {"error": {"type": "string"}}},
+    },
+)
+class BookmarkToggleView(APIView):
```
Bookmark 에 대해서 toggle 형태로 하나의 API 로 개발하는 방향이 맞을까요?
API 명세서에 따르면, 아래 2개로 멱등하게 분리되어 있는 걸로 되어있어서요
- Bookmark 추가 API 
- Bookmark 해제 API

Kochiya35 replied:
앗 수정하겠습니다

Meoyoug commented on apps/lectures/urls/v1_bookmarks.py:
```diff
@@ -0,0 +1,7 @@
+from apps.lectures.views.bookmark_views import BookmarkToggleView
+
+urlpatterns = [
+    path("bookmarks/toggle", BookmarkToggleView.as_view()),
```
여기 준식멘토님께서 말씀하신대로 엔드포인트가 두개로 나뉘어야할거같아요
```
api/v1/lectures/<int:lecture_id>/bookmarks/add
api/v1/lectures/<int:lecture_id>/bookmarks/cancel
```

이런식으로 나눠주세요

Meoyoug commented on apps/lectures/views/bookmark_views.py:
```diff
@@ -0,0 +1,55 @@
+class BookmarkToggleView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request):
```
위에 urls.py 커멘트대로 나누고 path param으로 강의 아이디 넘겨받아서 로직구성하시면 될 것 같습니다

Meoyoug reviewed (commented):
4번 마이그레이션 파일에 대해서 충돌나는데 확인후 수정부탁드립니다.

Umdoong commented on apps/lectures/views/bookmark_views.py:
```diff
@@ -0,0 +1,55 @@
+    },
+)
+class BookmarkToggleView(APIView):
+    permission_classes = [IsAuthenticated]
```
"DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.IsAuthenticatedOrReadOnly",),
이 코드로 인해서  post 요청은 default로 IsAuthenticated로 셋팅 되어서 필요 없는 코드 같습니다.

Kochiya35 committed:
💡 chore: config/urls, lecture/urls trailing slash 제거

config와 lecture의 url에서 마지막에 있는 슬래쉬를 제거했습니다.

# Conflicts:
#	config/urls.py

Kochiya35 committed:
✨ feat: bookmark togle에서 멱등성 있게 api분리

Kochiya35 committed:
💡 chore: lecture_id -> lecture_uuid로 변경 & 북마크 api로 분리돼있던 거 http method로 분리

lecture_id를 lecture_uuid로 변경, 북마크 추가 제거를 api로 분리했었는데 한 엔드포인트 내에서 http method로 분리하도록 변경

Kochiya35 committed:
🎨 style: black, isort 코드 포매터 실행

black, isort 코드 포매터 실행

Umdoong commented on apps/lectures/views/bookmark_views.py:
service와 repository 쪽을 활용해서 로직을 조금 더 분리해보시면 좋을 것 같습니다.

Umdoong commented:
mypy 쪽도 해결해주세요!!

Meoyoug reviewed (changes requested):
Test 터지고있는데 확인 후 수정부탁드립니다
아마 uuid를 pk로 오인해서 생긴거 같네요

Kochiya35 committed:
✨ feat:북마크 추가/삭제뷰 구현

북마크 추가/삭제기능을 궇년했늡니다

Kochiya35 committed:
✅ test: 테스트 파일 교체

북마크 API 테스트

Kochiya35 committed:
♻️ refactor : 리팩터링 (기능 변화 없음)

Kochiya35 committed:
🐛 fix:trailing slash 추가

경로에 trailing slash 를 추가해서 404를 해결

Meoyoug reviewed (approved):
고생하셧습니다 머지할게요~
